### ISSUE TITLE
bug --> Get Family Parameter

### ISSUE THREAD

**Issue by erfajo** [DynamoDS/Dynamo#9495]
## Dynamo version
2.0.2 --> any in the Dynamo 2.0.x range

## What did you do? 
Adding the node "Get Family Parameter" in a dynamo 2.0.x graph. On save is there a selected item in the pull-down. 

## What did you expect to see?
On reopening the graph I expect that the previously selected item in the pull-down was "remembered".

## What did you see instead?
The node does not save the selected index so the nodes are blank when reopening the graph
![capture](https://user-images.githubusercontent.com/2844374/52915268-10ed0300-32d2-11e9-85b3-2bf5b60e629a.png)

**Comment by QilongTang** [comment]
@erfajo Thank you for reporting, this is a regression we found after 2.X from Dynamo 1.3.4, it has been officially fixed in the coming release Dynamo 2.1.0, you can check a series of PR starting from https://github.com/DynamoDS/Dynamo/pull/9367. If you still has questions or concerns, feel free to comment and reopen.

**Comment by erfajo** [comment]
I have tried to see the code in the dynamo 2.1.x branch, but I dont think this solves that issue.
The code "Get Family Parameter" in has serialization/deserialization functions which parse the input port so the parameters can be found for that family. This works in the 1.3.x graphs since they are build on xml core. The 2.0.x graphs are moved to json. I cant see how this node gets its input value!?

however, the serialization/deserialization does not support the new way dropdowns are supported in the 1.3.x branch, but that is easy to  fix... this will not remove the wrong warning which also occur, but it respects the new "index" attribute.

```
#region Serialization/Deserialization
public static int GetIndex(string index)
{
    var t = -1;
    var splits = index.Split(':');
    if (splits.Count() > 1) { t = System.Convert.ToInt32(splits[0]); }
    else { t = System.Convert.ToInt32(index); }
    return t;
}

protected override void SerializeCore(XmlElement nodeElement, SaveContext context)
{
    base.SerializeCore(nodeElement, context);
    if (input != null) { nodeElement.SetAttribute("input", input); }
    else { nodeElement.SetAttribute("input", null); }
}

protected override void DeserializeCore(XmlElement nodeElement, SaveContext context)
{
    base.DeserializeCore(nodeElement, context);

    // get the stored index
    var index = nodeElement.Attributes["index"];
    if (index == null) return;
    int selectedIndex = GetIndex(index.Value);

    try
    {
        // get the stored input, set index
        var savedInput = nodeElement.Attributes["input"];
        if (savedInput.Value != "")
        {
            input = savedInput.Value;
            PopulateItems();
            SelectedIndex = selectedIndex;
        }
    }
    catch (System.Exception) { }
}
#endregion Serialization/Deserialization
```

**Comment by erfajo** [comment]
I can't reopen a closed issue, so it might be better not to close issues that fast... otherwise if I dont hear anything I have to make a new issue for giving this feedback... that is really silly.

**Comment by QilongTang** [comment]
@erfajo I would echo what I wrote in the PR, the new Json attribute we added in the node will try to "remember" what user selects when saving thus restore the selected Item when opening the file. I believe this is exactly the cause of the bug you reported. We have finished fixing and testing on our end and in DynamoForRevit, that is the reason I closed this issue with my response in the first place.

Here is an example for you to review as the new behavior:
```Json
"Nodes": [
    {
      "ConcreteType": "DSRevitNodesUI.Categories, DSRevitNodesUI",
      "SelectedIndex": 6,
      "SelectedString": "OST_DuctTerminal",
      "NodeType": "ExtensionNode",
      "Id": "e16b9ac4ad604e7eb4d29a7975335d7c",
      "Inputs": [],
      "Outputs": [
        {
          "Id": "ac2e1b04ef994a4fb5dbe4b2855d1097",
          "Name": "Category",
          "Description": "The selected Category.",
          "UsingDefaultValue": false,
          "Level": 2,
          "UseLevels": false,
          "KeepListStructure": false
        }
      ],
      "Replication": "Disabled",
      "Description": "All built-in categories."
    }
  ]
```

I don't think the goal here is to restore to the same index serializing/ deserilizing behavior back to v1.3.4 when index and string was mixed into. Since Json serialization/deserialization is based on class properties, the current way support the same behavior with slightly different information saved. Also since 2.1.0 is already internally released and consumed by Revit group, any changes to file format will be extremely hard for us. 

Let me know if you have more comments.

Cheers,
Aaron-

**Comment by erfajo** [comment]
This is not a DropDown issue as such, it is an input issue!
Where is the input stored?
the only thing you have in your json snippet is the dropdown part, there is no input... that node must have a famiily type or instance connected to the **F** input port

"Inputs": [],

![capture2 - kopi](https://user-images.githubusercontent.com/2844374/52964196-0ac65780-33a2-11e9-9362-221a25b22866.png)

**Comment by erfajo** [comment]
@mjkkirschner, @johnpierson it seems that @QilongTang didnt get what the problem was... could you find another one to assign the issue?

**Comment by QilongTang** [comment]
@erfajo Again, I would echo this is a drop down issue which has been fixed by me in v2.1.0. Please see the attached gif, I believe I did what you have described in the issue description and the bug can no longer be reproduced.

Cheers,
Aaron-
![getfamilyparameter](https://user-images.githubusercontent.com/3942418/53165632-3e230500-35a1-11e9-8b10-702a21cfdf66.gif)

**Comment by erfajo** [comment]
thanks, @QilongTang your demonstrations is convincing :-)
however, from a coding point of view have I lost the overview totally... in your snippet is there no indication of storing the input... I need to get my head around this so I can understand it :-)
